fms/models/gpt_bigcode.py

```diff
@@ -0,0 +1,322 @@
+                # In case of left-padding, prevent negative indices (get zeroed anyway)
+                _position_ids = _position_ids.clamp(min=0)
+
+            # zero out the associated position embeddings
+            position_out = self.position_embedding(_position_ids).mul(
+                ~is_pad.unsqueeze(-1)
+            )
+        else:
+            # just look up the position embeddings, no need to zero if no pad_id
+            position_out = self.position_embedding(_position_ids)
+
+        # perform absolute position embedding
+        x = x_emb + position_out
+
+        # apply dropout to embeddings
+        if self.config.emb_dropout:
+            x = self.emb_dropout(x)
+
+        # this is the output cache for all the decoder layers
+        present_key_value_states = []
+
+        for i, layer in enumerate(self.layers):
+            output = layer(
+                x=x,
+                mask=mask,
+                is_causal_mask=is_causal_mask,
+                past_key_value_state=past_key_value_states[i],
+                use_cache=use_cache,
+                attn_algorithm=attn_algorithm,
+            )
+
+            if use_cache:
+                x, present_key_value_state = output
+                present_key_value_states.append(present_key_value_state)
+
+            else:
+                x = output
+
+        dec_out = self.dec_norm(x)
+        if self.config.p_dropout:
+            dec_out = self.dropout(dec_out)
+
+        return dec_out, present_key_value_states
+
+
+# Implements the decoder-only GPTBigCodeModel
+class GPTBigCode(nn.Module):
+    def __init__(
+        self,
+        config: Optional[GPTBigCodeConfig] = None,
+        **kwargs,
+    ):
+        super(GPTBigCode, self).__init__()
+        if config is not None:
+            self.config = config
+        else:
+            self.config = GPTBigCodeConfig()
+        self.config: GPTBigCodeConfig = self.config.updated(**kwargs)
+
+        self.base_model = GPTBigCodeHeadless(self.config)
+        self.head = nn.Linear(
+            self.config.emb_dim, self.config.src_vocab_size, bias=self.config.vocab_bias
+        )
+
+        # this model ties weights, so we tie here
+        self.head.weight = self.base_model.embedding.weight
+
+        self.reset_params()
+
+    @classmethod
+    def from_config(cls, config: GPTBigCodeConfig) -> "GPTBigCode":
+        return cls(config)
+
+    def get_config(self) -> GPTBigCodeConfig:
+        return self.config
+
+    def reset_params(self):
+        # Modules are self-initializing, we're just going to down-scale the final prediction head to be
+        # mixed-fan (inputs and gradients scale to the same inverse factors) if it isn't tied
+        self.head.weight.data.normal_(
+            0,
+            1 / math.sqrt(math.sqrt(self.config.emb_dim * self.config.src_vocab_size)),
+        )
+
+    def forward(
+        self,
+        x: Optional[torch.LongTensor] = None,
```

Comment:
can we make this positional/required too (not `=None`).

Reply:
Yes, this should be required